### PR TITLE
feat: added rate limit

### PR DESCRIPTION
**Pull request by githubjeka**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #326 

http://localhost/api/user

image

http://localhost/api/user

> Too Many Requests

### CODE REVIEW

File: src/User/Controller/RateLimitToUserApiMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Yii\RateLimiter\Counter;
+use Yiisoft\Yii\RateLimiter\LimitRequestsMiddleware;
+use Yiisoft\Yii\RateLimiter\Policy\LimitPerIp;
+use Yiisoft\Yii\RateLimiter\Storage\StorageInterface;
+
+class RateLimitToUserApiMiddleware implements MiddlewareInterface
```
Comment (rustamwin) [inline src/User/Controller/RateLimitToUserApiMiddleware.php:17]: What is the benefit if it already exists?

Comment (githubjeka) [reply src/User/Controller/RateLimitToUserApiMiddleware.php:17]: It is compatible with desire for coherence and modeling Consistency. This is the only place where there is rate limiting logic for the user api.

Comment (githubjeka) [reply src/User/Controller/RateLimitToUserApiMiddleware.php:17]: This is not an ideal implementation and just demo code for subsequent implementations.

File: src/User/Controller/RateLimitToUserApiMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
```
[the remark this answers is not in the document]
Comment (githubjeka) [reply src/User/Controller/RateLimitToUserApiMiddleware.php]: done

File: config/web/rate-limit.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+use Yiisoft\Yii\RateLimiter\CounterInterface;
+use Yiisoft\Yii\RateLimiter\Storage\StorageInterface;
+use \Yiisoft\Yii\RateLimiter\Counter;
+
+/** @var array $params */
+
+return [
+    StorageInterface::class => function () {
+        $cache = new \Yiisoft\Cache\File\FileCache(__DIR__ . '/../runtime/rate-limit/');
```
Comment (xepozz) [inline config/web/rate-limit.php:13]: How about to use `Yiisoft\Aliases\Aliases` with `@runtime` value?
[Here](https://github.com/yiisoft/demo/blob/master/config/params.php#L52) you can see all aliases

File: config/routes.php [hunk 4]
```diff
@@ -42,12 +46,17 @@
...
     // Auth
     Route::methods([Method::GET, Method::POST], '/login')
+        ->middleware(LimitRequestsMiddleware::class)
         ->action([AuthController::class, 'login'])
         ->name('auth/login'),
     Route::post('/logout')
         ->action([AuthController::class, 'logout'])
         ->name('auth/logout'),
     Route::methods([Method::GET, Method::POST], '/signup')
+        ->middleware(fn(\Psr\Container\ContainerInterface $container) => new LimitRequestsMiddleware(
```
Comment (xepozz) [inline config/routes.php:56]: You can inject only needed objects without using the container. It's just the same, but a bit cleaner

Comment (xepozz) [inline config/routes.php:56]:
```suggestion
        ->middleware(fn(ResponseFactoryInterface $responseFactory, StorageInterface $storage) => new LimitRequestsMiddleware(
```

File: config/web/rate-limit.php (hunk 3 above)
Comment (githubjeka) [reply config/web/rate-limit.php:13]: [FileCache::__construct](https://github.com/yiisoft/cache-file/blob/master/src/FileCache.php#L105) does not support aliases

Comment (xepozz) [reply config/web/rate-limit.php:13]: And it should not.
[Here](https://github.com/yiisoft/demo/blob/de3b6fe1ae1d1a1429d5ed5f51aa467ec59a8714/config/common/translator.php#L15) is an example how to use `aliases`.
`$aliases->get('@runtime/rate-limiter')` === `/path/to/runtime/rate-limiter`

Comment (githubjeka) [reply config/web/rate-limit.php]: Thanks

**Comment by samdark** [comment]
👍

**Pull request merged**